src/Juvix/Core/IR/Typechecker/Env.hs

```diff
@@ -1,11 +1,25 @@
 module Juvix.Core.IR.Typechecker.Env where
 
+import Data.HashMap.Strict (HashMap)
 import Juvix.Core.IR.Typechecker.Log
 import Juvix.Core.IR.Typechecker.Types
-import Juvix.Library
+import qualified Juvix.Core.IR.Types as IR
+import Juvix.Library hiding (Datatype)
 
 data EnvCtx primTy primVal
-  = EnvCtx {typecheckerLog :: [Log primTy primVal]}
+  = EnvCtx
+      { typecheckerLog :: [Log primTy primVal],
+        globals :: Globals primTy primVal
+      }
+  deriving (Show, Eq, Generic)
+
+type Globals primTy primVal =
+  HashMap IR.GlobalName (Global primTy primVal)
+
+data Global primTy primVal
+  = GDatatype (IR.Datatype primTy primVal)
+  | GDataCon (IR.DataCon primTy primVal)
+  | GFunction (IR.Function primTy primVal)
```

Comment:
Is the idea that top-level functions will end up as `GFunction` ?

Reply:
Yeah, exactly, the intention is that a top level function (or a lifted one) is an element in the typechecker state of type `IR.Function`, and a datatype is one `Datatype` and zero or more `DataCon`s.

A datatype also has a list of all its constructors because this will presumably be necessary for the coverage checker etc